tremor-script/src/registry/custom_aggregate_fn.rs

```diff
@@ -0,0 +1,237 @@
+    /// # Errors
+    /// Returns an error if the initialization fails, or the init function does not return a value
+    pub fn init<'event>(&mut self, env: &Env<'_, 'event>) -> FResult<()>
+    where
+        'script: 'event,
+    {
+        let mut local_stack = LocalStack::with_size(128);
+
+        let mut no_meta = Value::null();
+        let mut state = Value::null().into_static();
+        let mut no_event = Value::null();
+
+        let mut body_iter = self.init_body.iter().peekable();
+        while let Some(expr) = body_iter.next() {
+            let env_local = Env {
+                context: env.context,
+                consts: env.consts,
+                aggrs: &[],
+                meta: env.meta,
+                recursion_limit: env.recursion_limit,
+            };
+
+            let cont = expr.run(
+                ExecOpts {
+                    result_needed: true,
+                    aggr: AggrType::Tick,
+                },
+                &env_local,
+                &mut no_event,
+                &mut state,
+                &mut no_meta,
+                &mut local_stack,
+            )?;
+
+            if body_iter.peek().is_none() {
+                if let Cont::Cont(value) = cont {
+                    self.state = value.into_owned().clone_static();
+                } else {
+                    return Err(FunctionError::NoAggregateValueInInit);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Aggregate a value
+    /// # Errors
+    /// Will return an error if the aggregate function does not return a value, of it errors
+    pub fn aggregate<'event>(&mut self, args: &[&Value], env: &Env<'_, 'event>) -> FResult<()>
+    where
+        'script: 'event,
+    {
+        let mut body_iter = self.aggregate_body.iter().peekable();
+        let mut no_meta = Value::null();
+        let mut state = Value::null().into_static();
+        let mut no_event = Value::null();
+        let mut local_stack = LocalStack::with_size(2);
+        local_stack.values.insert(0, Some(self.state.clone()));
+        local_stack
+            .values
+            .insert(1, args.get(0).map(|x| (*x).clone()));
+
+        let env = Env {
+            context: env.context,
+            consts: env.consts,
+            aggrs: &[],
+            meta: env.meta,
+            recursion_limit: env.recursion_limit,
+        };
+
+        while let Some(expr) = body_iter.next() {
+            let cont = expr.run(
+                ExecOpts {
+                    result_needed: true,
+                    aggr: AggrType::Tick,
+                },
+                &env,
+                &mut no_event,
+                &mut state,
+                &mut no_meta,
+                &mut local_stack,
+            )?;
+
+            if body_iter.peek().is_none() {
+                if let Cont::Cont(value) = cont {
+                    self.state = value.into_owned().clone_static();
```

Comment:
I don't think we need `into_owned()` and `clone_static()` here as it would clone the value twice